File: nreduce/benchmark/bintree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct bintree {
  struct bintree *left;
  struct bintree *right;
} bintree;

int countnodes(bintree *tree)
{
  if (tree)
    return 1 + countnodes(tree->left) + countnodes(tree->right);
  else
    return 0;
}

bintree *mktree(int depth, int max)
{
  if (depth == max) {
    return NULL;
  }
  else {
    bintree *tree = (bintree*)malloc(sizeof(bintree));
    tree->left = mktree(depth+1,max);
    tree->right = mktree(depth+1,max);
    return tree;
  }
}

void freetree(bintree *tree)
{
  if (tree) {
    freetree(tree->left);
    freetree(tree->right);
    free(tree);
  }
}

int main(int argc, char **argv)
{
  bintree *tree;
  int n = 16;
  if (2 <= argc)
    n = atoi(argv[1]);

  tree = mktree(0,n);
  printf("%d\n",countnodes(tree));
  freetree(tree);
  return 0;
}

File: nreduce/benchmark/bintree.js
function node(left,right)
{
  this.left = left;
  this.right = right;
}

function bintree(depth,max)
{
  if (depth == max)
    return null;
  else
    return new node(bintree(depth+1,max),bintree(depth+1,max));
}

function countnodes(tree)
{
  if (tree)
    return 1 + countnodes(tree.left) + countnodes(tree.right);
  else
    return 0;
}

var n = 16;
if (arguments.length > 0)
  n = arguments[0];

var tree = bintree(0,n);
var count = countnodes(tree);
print(count);
